### PR TITLE
impl(generator): emit a proto export file for discovery defined services

### PR DESCRIPTION
**Pull request by scotthart**
Compute users who leverage IWYU in their build system are experiencing issues with relying on transitive includes for internal/common_xxx.pb.h files. This PR ameliorates this issue by providing an additional header file to include, per service, that exports the common.pb.h files that service.proto requires, satisfying IWYU.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819)

### CODE REVIEW

File: generator/internal/discovery_proto_export_file.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+// Creates a IWYU pragma export file of the internal protos needed by a service.
+class DiscoveryProtoExportFile {
+ public:
+  DiscoveryProtoExportFile() = default;
```
Comment (devbww) [inline generator/internal/discovery_proto_export_file.h:29]: It looks like the default ctor isn't used.

File: generator/internal/discovery_to_proto_test.cc [hunk 2]
```diff
@@ -1651,16 +1651,22 @@ TEST_F(AssignResourcesAndTypesToFilesTest,
...
-  ASSERT_THAT((*result)[1].types().size(), Eq(1));
-  EXPECT_THAT((*result)[1].types().front()->name(), Eq("Operation"));
+  ASSERT_THAT(result->first[1].types().size(), Eq(1));
+  EXPECT_THAT(result->first[1].types().front()->name(), Eq("Operation"));
+
+  ASSERT_THAT(result->second, SizeIs(1));
+  EXPECT_THAT(result->second[0].relative_file_path(),
+              Eq("google/cloud/product_name/foos/version/foos_proto_export.h"));
+  //  EXPECT_THAT(result->second[0].proto_includes(),
+  //  ElementsAre("google/cloud/product_name/version/internal/common_000.proto"));
```
Comment (devbww) [inline generator/internal/discovery_to_proto_test.cc:1669]: Commented-out code?

File: generator/internal/discovery_proto_export_file.cc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+// uses IWYU pragmas to export the internal/common_xxx.pb.h files as an
+// IWYU-friendly alternative.
+
+#ifndef $header_include_guard$
+#define $header_include_guard$
+
+)""");
+  printer.Print("// IWYU pragma: begin_exports\n");
+  for (auto const& p : proto_includes_) {
+    printer.Print(absl::StrCat("#include <",
```
Comment (devbww) [inline generator/internal/discovery_proto_export_file.cc:75]: You could avoid the `StrCat()` by using separate `Print()` calls.

File: generator/internal/discovery_proto_export_file_test.cc [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+namespace {
+
+using ::testing::Eq;
+
+TEST(DiscoveryProtoExportFileTest, FormatFile) {
+  auto constexpr kExpectedHeaderFile = R"""(// Copyright 2023 Google LLC
```
Comment (devbww) [inline generator/internal/discovery_proto_export_file_test.cc:27]: This looks like a time bomb.

File: generator/internal/discovery_proto_export_file_test.cc [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+// IWYU pragma: begin_exports
+#include <google/cloud/compute/v1/internal/common_001.pb.h>
+#include <google/cloud/compute/v1/internal/common_002.pb.h>
+// IWYU pragma: end_exports
+
+#endif  // GOOGLE_CLOUD_CPP_RELATIVE_FILE_PATH
+)""";
+
+  DiscoveryProtoExportFile f(
+      "my_output_file_path", "relative_file_path",
```
Comment (devbww) [inline generator/internal/discovery_proto_export_file_test.cc:63]: _"relative_file_path.h"_ might be nicer.

File: generator/internal/discovery_proto_export_file_test.cc [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+}  // namespace
+}  // namespace generator_internal
+}  // namespace cloud
+}  // namespace google
+
+int main(int argc, char** argv) {
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
Comment (devbww) [inline generator/internal/discovery_proto_export_file_test.cc:80]: I don't know why the "google_cloud_cpp_generator_unit_tests" don't use `gtest_main`.

File: generator/internal/discovery_to_proto_test.cc [hunk 7]
```diff
@@ -1651,16 +1651,22 @@ TEST_F(AssignResourcesAndTypesToFilesTest,
...
+  EXPECT_THAT(result->first[0].types(), IsEmpty());
+  EXPECT_THAT(result->first[1].file_path(),
               Eq("output_path/google/cloud/product_name/"
                  "version/internal/common_000.proto"));
-  ASSERT_THAT((*result)[1].types().size(), Eq(1));
-  EXPECT_THAT((*result)[1].types().front()->name(), Eq("Operation"));
+  ASSERT_THAT(result->first[1].types().size(), Eq(1));
+  EXPECT_THAT(result->first[1].types().front()->name(), Eq("Operation"));
+
+  ASSERT_THAT(result->second, SizeIs(1));
```
Comment (devbww) [inline generator/internal/discovery_to_proto_test.cc:1665]: Use `ElementsAre()` to avoid the need for size assertions before testing values?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819)* status: 0 of 105 files reviewed, 10 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/codegen_utils.cc` line 366 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346711087:-Ng-cUDX3_m4Hc_OJdbK:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/codegen_utils.cc#L366)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Observation: If you do the `StrCat()` first, you could use the in-place versions of `AsciiStrToUpper()` and `StrReplaceAll()`, and save a couple of allocations/deallocations/copies.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_proto_export_file.h` line 29 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346724728:-Ng-cVNq5vry-TFe0SNH:bss4a2b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file.h#L29)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like the default ctor isn't used.
</blockquote></details>

Removed.
___
*[`generator/internal/discovery_proto_export_file.cc` line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346716449:-Ng-cXRn9rcYovPifsTK:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file.cc#L57)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

You should be able to drop the `.c_str()` now that `Print()` takes a `string_view`.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_proto_export_file.cc` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346728228:-Ng-cZbHC9j05htVOxLp:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file.cc#L75)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

You could avoid the `StrCat()` by using separate `Print()` calls.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_proto_export_file.cc` line 78 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346716903:-Ng-cZEu100B8RP97JyQ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file.cc#L78)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Ditto on `.c_str()`.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_proto_export_file_test.cc` line 27 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346730359:-Ng-dqDO6OB16Dwir4Ez:bxg1m45) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file_test.cc#L27)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This looks like a time bomb.
</blockquote></details>

Made year variable.
___
*[`generator/internal/discovery_proto_export_file_test.cc` line 63 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346736217:-Ng-caDS4cdOqdWv3j_M:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file_test.cc#L63)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

_"relative_file_path.h"_ might be nicer.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_proto_export_file_test.cc` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346736664:-Ng-caXQ9iAqDGIIzlrK:bc8zb1x) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file_test.cc#L80)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I don't know why the "google_cloud_cpp_generator_unit_tests" don't use `gtest_main`.
</blockquote></details>

because the unit tests link libprotoc and libprotoc defines a `main` symbol that depending on linkage order sometimes get used instead of `gtest_main`. this is merely a workaround until protobuf fixes libprotoc
___
*[`generator/internal/discovery_to_proto_test.cc` line 1665 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346740296:-Ng-dwzM-JWD3478nYKn:b-619ne4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_to_proto_test.cc#L1665)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Use `ElementsAre()` to avoid the need for size assertions before testing values?
</blockquote></details>

In this case it would require writing a custom matcher that would only be used once.
___
*[`generator/internal/discovery_to_proto_test.cc` line 1669 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346725896:-Ng-dw37Aqa-Mrlq1FBU:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_to_proto_test.cc#L1669)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Commented-out code?
</blockquote></details>

Fixed.

**Review by devbww — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819)* status: 0 of 105 files reviewed, 3 unresolved discussions (waiting on @coryan and @scotthart)

___
*[`generator/internal/codegen_utils.cc` line 366 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346711087:-Ng-g6DH1wBmdMiopwo8:b6no8f) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/codegen_utils.cc#L366)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

That still uses the value-returning versions (e.g., `std::string AsciiStrToUpper(absl::string_view)` instead of `void AsciiStrToUpper(std::string*)`), but no big deal.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819)* status: 0 of 105 files reviewed, 3 unresolved discussions (waiting on @coryan and @devbww)

___
*[`generator/internal/discovery_proto_export_file_test.cc` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12819#gh-1346736664:-Ng-ixcxExVletcfWq6D:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2a955f370a1d330d54a4882d4981806caa7e5bb6/generator/internal/discovery_proto_export_file_test.cc#L80)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This has come up before, if it is not too late, maybe add a comment?
</blockquote></details>

Done.

**Pull request merged**